### DISCUSSION TITLE
How to train a single component

### DISCUSSION THREAD

**Discussion by oliviercwa** [explosion/spaCy#8988] [Help: Coding & Implementations] [answered]
Hello,
I am using spacy 3.1.1 and I want to train a single custom component. I created my pipeline using en_core_web_sm_3.1.0 by
1- Loading the config from en_core_web_sm_3.1.0
2- Then for each of the following component: [**tok2vec, tagger, attribute_ruler, parser, lemmatizer**]:
     a) - I load the pipe config using _get_configs_ below (code borrowed from language.prx)
```python
def get_configs(config: Config, pipe_name: str):
    filled = config
    interpolated = filled.interpolate() if not filled.is_interpolated else filled
    pipeline = interpolated.get("components", {})
    pipe_cfg = util.copy_config(pipeline[pipe_name])
    pipe_cfg.pop("factory")
    raw_config = Config(filled["components"][pipe_name])
    return pipe_cfg, raw_config
```

b) - call add_pipe
c) - load from disk the model

This works very well when running a model (e.g. nlp('this is a test'))

**Now I am adding a custom component at the end and I just want to train this custom component**, not the other ones.
I created the set of examples, and I call 
```python
nlp = spacy.blank('en')
...
# build the pipeline as described above
...
nlp.initialize(lambda: examples)
```

Two questions:
1- it seems that the pipeline expect to train all components (lemmatizer, etc.). At least initialize is called on every component when I just want to train my custom component leveraging of course the tokenizer, lemmatizer and parser. Is there a way in the config to mention that only one component should be train ? 

2- I am getting the following error
```Can't find table(s) lemma_rules for language 'en' in spacy-lookups-data. Make sure you have the package installed or provide your own lookup tables if no default lookups are available for your language.```

Am I missing something ?

Thank you

**Comment by adrianeboyd** [comment, marked as answer]
Hi, I think what you're looking for is called "frozen components", see the docs: https://spacy.io/usage/training#config-custom

If you need to run some components in the pipeline to provide input for the component you're training (like NER or sentence boundaries), but without training/updating those components themselves, use "annotating components": https://spacy.io/usage/training#annotating-components

You can have components that are both frozen and annotating.

**Reply by oliviercwa** [reply]
Thank you very much Adriane. This seems to be what I was looking for!

**Comment by oliviercwa** [comment]
@adrianeboyd ,
I have noticed that the only way to provide the "frozen components" via code (not via CLI) is to call spacy.training.initialize.init_nlp directly. In other words, the code in my original question never sets the frozen components.

Also I am still getting the error from spacy\lookups.py line 30. The same config works great if I build the pipeline and use it to evaluate rather than to train

```
[E955] Can't find table(s) lexeme_norm for language 'en' in spacy-lookups-data. Make sure you have the package installed or provide your own lookup tables if no default lookups are available for your language.
  File "\external\spaCy\spacy\lookups.py", line 30, in load_lookups
    raise ValueError(Errors.E955.format(table=", ".join(tables), lang=lang))
  File "\external\spaCy\spacy\language.py", line 98, in load_lookups_data
    lookups = load_lookups(lang=lang, tables=tables)
  File "\external\spaCy\spacy\language.py", line 1246, in initialize
    I = registry.resolve(config["initialize"], schema=ConfigSchemaInit)
  File "\external\spaCy\spacy\training\initialize.py", line 82, in init_nlp
    nlp.initialize(lambda: train_corpus(nlp), sgd=optimizer)
```

What I am missing ?

Thanks

**Reply by adrianeboyd** [reply]
Did you install `spacy-lookups-data`? Alternatively you can provide `lexeme_norm` from an a custom source or remove it from the config.

**Reply by oliviercwa** [reply]
In my config, I have this defined. And the model is a variation on en_core_web_sm.
Is more needed ?
```
[initialize.lookups]
@misc = "spacy.LookupsDataLoader.v1"
lang = ${nlp.lang}
tables = ["lexeme_norm"]
```

**Reply by adrianeboyd** [reply]
You need to have `spacy-lookups-data` installed for `spacy.LookupsDataLoader.v1` to work. Is that your question?

**Reply by oliviercwa** [reply]
Here is a sample that illustrates the problem. Without going into details, there are good reasons why I am building the pipeline this way.
```python
# Create pipeline from config
ner_root_path: Path = Path(distutils.sysconfig.get_python_lib()) / \
                           'en_core_web_sm' / \
                           'en_core_web_sm-3.1.0'
config = util.load_config(ner_root_path / 'config.cfg')
nlp = spacy.blank(config['nlp']['lang'], config=config)
for pipe_name in nlp.pipe_names:
  pipe = nlp.get_pipe(pipe_name)
  pipe.from_disk(ner_root_path / pipe_name, exclude=['vocab'])

# this works
predicted = nlp("This is a yogurt")
reference = nlp("This is a yogurt")
ent1 = reference[3:4]
ent1.label_ = 'FOOD'
reference.ents = [ent1]

# Training
examples = []
examples.append(Example(predicted=predicted, reference=reference))
```

As result I am getting ```Can't find table(s) lexeme_norm for language 'en' in spacy-lookups-data```

**Reply by oliviercwa** [reply]
Ho I am so sorry, I did not realize it was a separate installation ! Trying it now. Thanks
